src/eos-updater-apply.c

```diff
@@ -73,41 +77,592 @@ get_test_osname (void)
+        }
+    }
+
+  return TRUE;
+}
+
+static const gchar *
+get_pending_flatpak_deployments_state_path (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_UPDATER_FLATPAK_UPGRADE_STATE_DIR",
+                                    LOCALSTATEDIR "/lib/eos-application-tools");
+}
+
+/* We return a transfer-full GFile of the record of flatpaks to be
+ * deployed on the next boot. If something goes wrong the caller is expected
+ * to delete the file. In any event, the caller must free the return value. */
+static GFile *
+write_flatpaks_to_be_deployed (GPtrArray    *flatpaks_to_deploy,
+                               const gchar  *deployment_csum,
+                               GCancellable *cancellable,
+                               GError       **error)
+{
+  const gchar *pending_flatpak_deployments_state_path = get_pending_flatpak_deployments_state_path ();
+  g_autofree gchar *pending_flatpak_deployments_to_export_dir_path = g_build_filename (pending_flatpak_deployments_state_path,
+                                                                                       deployment_csum,
+                                                                                       NULL);
+  g_autofree gchar *pending_flatpak_deployments_path = g_build_filename (pending_flatpak_deployments_to_export_dir_path,
+                                                                         "to-export",
+                                                                         NULL);
+  g_autoptr(GFile) pending_flatpak_deployments_to_export_dir = g_file_new_for_path (pending_flatpak_deployments_to_export_dir_path);
+  g_auto(GStrv) flatpak_names_to_autodeploy = g_new0 (gchar *, flatpaks_to_deploy->len + 1);
+  g_autofree gchar *flatpak_names_to_autodeploy_contents = NULL;
+  g_autoptr(GError) local_error = NULL;
+  gsize i = 0;
+
+  for (; i < flatpaks_to_deploy->len; ++i)
+    {
+      FlatpakRemoteRef *to_install = (FlatpakRemoteRef *) (g_ptr_array_index (flatpaks_to_deploy, i));
+      g_autofree gchar *flatpak_ref = flatpak_ref_format_ref (FLATPAK_REF (to_install));
+      flatpak_names_to_autodeploy[i] = g_strjoin(":",
+                                                 flatpak_remote_ref_get_remote_name (to_install),
+                                                 flatpak_ref,
+                                                 NULL);
+    }
+
+  flatpak_names_to_autodeploy_contents = g_strjoinv ("\n", flatpak_names_to_autodeploy);
+
+  /* Directory containing the flatpak deployments path is not guaranteed
+   * to exist, so create it first */
+  if (!g_file_make_directory_with_parents (pending_flatpak_deployments_to_export_dir,
+                                           cancellable,
+                                           &local_error))
+    {
+      if (!g_error_matches (local_error, G_FILE_ERROR, G_FILE_ERROR_EXIST))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return NULL;
+        }
```

Comment:
Need to `g_clear_error(&local_error)` here to avoid confusion if the variable is used again later when this function is modified.